flask/app/gene_viewer.py

```diff
@@ -0,0 +1,250 @@
+
+    filters = [models.Gene.gene == gene for gene in genes]
+
+    if user_id:
+        app.logger.debug("Processing query - restricted based on user id.")
+        query = (
+            models.Gene.query.join(models.Gene.variants)
+            .join(models.Variant.genotype)
+            .join(models.Genotype.analysis, models.Genotype.dataset)
+            .join(models.Dataset.tissue_sample)
+            .join(models.TissueSample.participant)
+            .join(models.Participant.family)
+            .join(
+                models.groups_datasets_table,
+                models.Dataset.dataset_id
+                == models.groups_datasets_table.columns.dataset_id,
+            )
+            .join(
+                models.users_groups_table,
+                models.groups_datasets_table.columns.group_id
+                == models.users_groups_table.columns.group_id,
+            )
+            .options(
+                contains_eager(models.Gene.variants)
+                .contains_eager(models.Variant.genotype)
+                .contains_eager(models.Genotype.analysis)
+                .contains_eager(models.Analysis.datasets)
+                .contains_eager(models.Dataset.tissue_sample)
+                .contains_eager(models.TissueSample.participant)
+                .contains_eager(models.Participant.family),
+            )
+            .filter(or_(*filters), models.users_groups_table.columns.user_id == user_id)
+        )
+
+    else:
+        app.logger.debug("Processing query - unrestricted based on user id.")
+        query = (
+            models.Gene.query.join(models.Gene.variants)
+            .join(models.Variant.genotype)
+            .join(models.Genotype.analysis, models.Genotype.dataset)
+            .join(models.Dataset.tissue_sample)
+            .join(models.TissueSample.participant)
+            .join(models.Participant.family)
+            .options(
+                contains_eager(models.Gene.variants)
+                .contains_eager(models.Variant.genotype)
+                .contains_eager(models.Genotype.analysis)
+                .contains_eager(models.Analysis.datasets)
+                .contains_eager(models.Dataset.tissue_sample)
+                .contains_eager(models.TissueSample.participant)
+                .contains_eager(models.Participant.family),
+            )
+            .filter(or_(*filters))
+        )
+
+    # defaults to json unless otherwise specified
+    app.logger.info(request.accept_mimetypes)
+
+    if request.accept_mimetypes["application/json"]:
+
+        app.logger.info("Defaulting to json response")
+
+        q = query.all()
+
+        if len(q) == 0:
+            app.logger.error("No requested genes were found.")
+            abort(400, description="No requested genes were found.")
+        elif len(q) < len(genes):
+            app.logger.error("Not all requested genes were found.")
+            abort(400, description="Not all requested genes were found.")
```

Comment:
Is the intent here to catch any invalid gene names? This also fails if the gene name is valid but we don't have any variants stored for that gene.

Reply:
Yes that was the intent - this was assuming that the only genes populated in the gene table would be found in the reports (and would therefore mean a gene has at least one variant) but this obviously doesn't work if the table is pre-populated.